tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner_test.py

```diff
@@ -0,0 +1,577 @@
+      'after': tf.zeros_like(weight, dtype=weight.dtype.base_dtype) * -1,
+      'sparsity': tf.constant(-1, dtype=tf.float64),
+      'before_sparsity': tf.constant(-1, dtype=tf.float64)
+      }
+      grad1 = tf.constant([67., 45., 89., 56., 100., 34., 1., 23., 12., 78.])
+      grad1 = tf.cast(grad1, self.grad(weight, 10).dtype)
+      grad2 = tf.constant([45., 78., 89., 34., 100., 23., 56., 67., 1., 12.])
+      grad2 = tf.cast(grad2, self.grad(weight, 10).dtype)
+      for i in tf.range(0, 5):
+        mask_before_update = optimizer.get_slot(weight, 'mask').read_value()
+        mask_before_update3 = optimizer.get_slot(weight, 'mask').read_value()
+        before_sparsity = tf.math.count_nonzero(mask_before_update, dtype=tf.int32) / tf.size(mask_before_update)
+        grad = p.preprocess_weights(optimizer, weight, grad1) 
+        p.postprocess_weights(optimizer, weight, grad)
+        if i == 1:
+          mask_after_update = optimizer.get_slot(weight, 'mask').read_value()
+          after_sparsity = tf.math.count_nonzero(mask_after_update, dtype=tf.int32) / tf.size(mask_before_update)
+          expected_iter1 = {'before': mask_before_update, 'after': mask_after_update, 
+                            'sparsity': after_sparsity, 'before_sparsity': before_sparsity}
+          grad1 = grad2
+          weight3 = tf.constant([3., 4., 0.1, 0.3, 5., 9., 10., 2., 1., 8.])
+          weight.assign(tf.math.add(weight3, sample_noise(i, dtype=weight.dtype)))
+        elif i == 3:
+          mask_after_update = optimizer.get_slot(weight, 'mask').read_value()
+          after_sparsity = tf.math.count_nonzero(mask_after_update, dtype=tf.int32) / tf.size(mask_before_update3)
+          expected_iter3 = {'before': mask_before_update3, 'after': mask_after_update, 
+                            'sparsity': after_sparsity, 'before_sparsity': before_sparsity}
+        optimizer.iterations.assign_add(1)
+      return expected_iter1, expected_iter3
+
+    result1, result2 = _train(optimizer, weight)
+    expected1 = tf.constant([0., 0., 1., 1., 1., 0., 1., 1., 0., 0.])
+    expected2 = tf.constant([0., 1., 1., 0., 1., 1., 1., 0., 0., 0.])
+    self.assertAllEqual(result1['sparsity'], self.target_sparsity)
+    self.assertAllEqual(result2['sparsity'], self.target_sparsity)
+    # # assert that sparsity does not change
+    self.assertAllEqual(tf.reduce_sum(result1['before']), tf.reduce_sum(result1['after']))
+    self.assertAllEqual(tf.reduce_sum(result2['before']), tf.reduce_sum(result2['after']))
+    # # assert there are some changes in the mask during each update
+    self.assertAllEqual(result1['after'], result2['before'])
+    self.assertAllEqual(expected1, result1['after'])
+    self.assertAllEqual(expected2, result2['after'])
+    del result1, result2
+    del expected1, expected2, mask, optimizer
+
+
+    weight1 = tf.Variable(np.linspace(1.0, 100.0, 10), dtype=tf.float32)
```

Comment:
Seems like duplicated code, maybe parameterize the test method w/ a boolean `use_function` and if true you tf.function the train method?